controller.py

```diff
@@ -9,80 +9,107 @@ def __init__(self,path):
         """Display criteria information"""
         self.view.view_criteria(self.shortlist.role,self.shortlist.role.criteria)
-        self.current_view = "criteria"
 
-    def show_role_info(self):
+    def show_role_info(self,k=None):
         """Display role information"""
         self.view.view_role(self.shortlist.role)
-        self.current_view = "role"
 
-    def show_applicants_list(self):
+    def show_applicants_list(self,k=None):
         """List all applicants"""
         self.view.view_applicants_list(self.shortlist)
-        self.current_view = "applicants_list"
+        self.options = self.options_applist
 
-    def show_applicant_details(self):
+    def show_applicant_details(self,k=None):
         """Select an applicant and view details"""
         try:
             i = int(input("Please enter the applicant number:"))
 
             self.current_applicant = self.shortlist.applicants[i-1]
             self.view.view_applicant_details(self.current_applicant)
-            self.current_view = "applicant_details"
+            self.options = self.options_appdetail
         except (ValueError, IndexError):
             pass
 
-    def open_applicant_pdf(self):
+    def open_applicant_pdf(self,k=None):
         """Open current applicant's CV"""
         startfile(self.current_applicant.cv)
 
+    def edit_appscore_start(self,k=None):
+        """select a criteria to edit score for"""
+        self.view.view_criteria(self.shortlist.role,self.shortlist.role.criteria)
+        options = [str(i) for i in range(len(self.shortlist.role.criteria))]
+        self.options = {i:self.edit_criteria_select for i in options} #generate options based on how many criteria there are
+        #if a criteria number is selected, call the next function
+
+    def edit_criteria_select(self, k=None):
+        self.current_criterion = self.shortlist.role.criteria[int(k)]
+        print(f"You selected {self.current_criterion.name}. Select the score you want to edit")
+        options = [str(i) for i in range(len(self.current_criterion.scores))]
+        self.options = {i:self.edit_score_select for i in options}
+        
+        for index,score in enumerate(self.current_criterion.scores):
+                print(f"{index}: {score}")
+        
+        self.current_score = self.current_criterion.scores[int(k)]
+
+    def edit_score_select(self, k=None):
+        print(f"You selected to update: {self.current_criterion.name} with the score of {self.current_score}")
+        print("press 1 to continue with this change, or 0 to exit editing")
+        
+        self.options = {"1":self.edit_score_confirm,
+                        "0":self.edit_criteria_quit}
+              
+    def edit_score_confirm(self,k=None):
+        """Confirm changes to score"""
+        print(f"Updated score: {self.current_criterion.name}:{self.current_score} and back to (applicant details)...")
+        self.current_applicant.scores.update({self.current_criterion.name:self.current_score})
```

Comment:
1. The key should be the criterion instance, not the name
2. You can use simpler dictionary assignment syntax here: `self.current_applicant.scores[self.current_criterion] = self.current_score`

Reply:
However, when I try to save the Criterion instance as the key, I get TypeError: unhashable type: "Criterion"

Reply:
Ah, okay. You'll learn what that means soon enough. In the meantime, dataclasses make it easy to make a dataclass hashable (and immutable). Update the decorator on the Criterion class to `@dataclass(frozen=True)`

Reply:
Adding (frozen = True) returns a different TypeError instead:  
![image](https://github.com/user-attachments/assets/4a3fe63f-ec7b-4c7f-9827-6327a3d20516)

It seems that it's because the score attribute in the Criterion dataclass is still considered mutable.

Reply:
tuples are an immutable, ordered, type in Python which you could use to store the scores for a Criterion instead of a list.